cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -1615,6 +1616,10 @@ public String generatePerBrokerConfiguration(int nodeId, Map<Integer, Map<String
      * @return  String with the Kafka broker configuration
      */
     private String generatePerBrokerConfiguration(NodeRef node, KafkaPool pool, Map<Integer, Map<String, String>> advertisedHostnames, Map<Integer, Map<String, String>> advertisedPorts)   {
+        // When doing manual rolling updates, empty maps are passed for advertised listeners and allowReconfiguration set to false.
+        // In broker case, we do not generate configurations if allowReconfiguration is set to false therefore this functions is never called.
+        // However, for controllers, we always generate configuration to read a controller specific configuration to use it for quorum healthcheck.
+        // Therefore, advertised hostname and ports should be nullable if empty maps are passed.
```

Comment:
I don't think you should generate a new configuration here. You have to roll it with the old configuration since otherwise it might not start because of other missing parts. So this seems wrong.

Reply:
This function only generates the configurations (including user defined configurations from the CR), but does not actually applies them. We pass this function for retrieving the desired configurations. For controllers, once we read the configurations, we don't actually update the existing configurations. For normal brokers, we pass this function but never execute it for manual rolling update which is the only place that sets advertised hostnames and ports to empty maps. I will try improving the comment perhaps.

Reply:
So, are you saying that you have to generate the configuration for controllers just to get the _tolerated lag_ to check if the controllers are up to date?

Reply:
The alternative we could do is, creating a separate function (for example, called getUserDefinedConfigurations) that only manual rolling update pass for desired configs. This function would only read user configs using `KafkaBrokerConfigurationBuilder.withUserConfiguration` without any listener or other config information as that's all the controller needs. This would be fine for normal brokers as this function does not get executed anyway when doing manual rolling update as `allowReconfiguration`  is set to false. What do people think?

Reply:
Actually, looking at the code little bit more, we don't even need `KafkaBrokerConfigurationBuilder.withUserConfiguration` because all it does is basically return `kafka.getConfiguration().getConfiguration()`. We could do this directly in the maybeRollKafka() function which would make it much simpler. So it would look like this:
```
protected Future<Void> maybeRollKafka(
            Set<NodeRef> nodes,
            Function<Pod, RestartReasons> podNeedsRestart,
            Map<Integer, Map<String, String>> kafkaAdvertisedHostnames,
            Map<Integer, Map<String, String>> kafkaAdvertisedPorts,
            boolean allowReconfiguration
    ) {
        return ReconcilerUtils.clientSecrets(reconciliation, secretOperator)
                .compose(compositeFuture ->
                        new KafkaRoller(
                                reconciliation,
                                vertx,
                                podOperator,
                                1_000,
                                operationTimeoutMs,
                                () -> new BackOff(250, 2, 10),
                                nodes,
                                compositeFuture.resultAt(0),
                                compositeFuture.resultAt(1),
                                adminClientProvider,
                                allowReconfiguration ?
                                        brokerId -> kafka.generatePerBrokerConfiguration(brokerId, kafkaAdvertisedHostnames, kafkaAdvertisedPorts) :
                                        brokerId -> kafka.getConfiguration().getConfiguration(),
                                logging,
                                kafka.getKafkaVersion(),
                                allowReconfiguration,
                                eventsPublisher
                        ).rollingRestart(podNeedsRestart));
    }
```

Reply:
I have pushed the proposed change above. Please let us know if you think this should be done differently.

Reply:
Based on the discussion below, we provide dummy advertised listeners maps instead of empty maps from manual rolling function.